### PR TITLE
Rewrite the wallpaper API

### PR DESCRIPTION
**Pull request by Elv13**
After sitting incomplete for many years, here is a ~~minimally working draft of the~~ new wallpaper API. The goal of the rewrite is to turn the wallpapers into a non-interactive wibox-like object. By doing this, it gets the full widget API to compose the scene. First of all, this is much more in line with other APIs. It is also easy to document. Last, but not least, it offers a path forward for SVG and HiDPI corner cases which the old would not handle. Could the old API be extended to support the full list of SVG+DPI+stretch corner cases? Technically yes, but the codebase would more than triple in size and be very redundant across the maximized/centered/tiled variants.

Before:

```lua

-- The old code used imperative code. It was one of the last place in `rc.lua`
-- which had imperative code. There was a long ongoing effort to get rid of
-- all imperative code to improve the learning curve.
screen.connect_signal("request::wallpaper", function(s)
    -- Wallpaper
    if beautiful.wallpaper then
        local wallpaper = beautiful.wallpaper
        -- If wallpaper is a function, call it with the screen
        if type(wallpaper) == "function" then
            wallpaper = wallpaper(s)
        end
        gears.wallpaper.maximized(wallpaper, s, true)
    end
end)
```

After:

```lua
-- Note that it doesn't *need* the tiled container, but it's convenient to add it to rc.lua.
-- By changing some booleans, you can turn can switch from tiled/centered/maximized
-- with no extra code.
screen.connect_signal("request::wallpaper", function(s)
    awful.wallpaper {
        screen = s,
        widget = {
            {
                image     = beautiful.wallpaper,
                upscale   = true,
                downscale = true,
                widget    = wibox.widget.imagebox,
            },
            valign = "center",
            halign = "center",
            tiled  = false,
            widget = wibox.container.tile,
        }
    }
end)
```

- [x]  multiscreen with single background widget (joint and disjoint) wibox.container.view?
- [x]  test gradient panning across multi-screen
- [x]   test image panning screen multi-screen
- [x]  test solid color
- [x]  test gradient
- [x]  test multi-screen
- [x]  test reolution changes
- [x]  test screen added
- [x]  test DPI changes
- [x]  test tiled
- [x]  test centered/aligned
- [x]  test stretch
- [x]  test resize (no stretch)
- [x]  test SVG stretch
- [x]  test SVG centered (with multiple DPIs)
- [x]  test honor_workarea
- [x]  test honor_padding
- [x]  test new awful.wallpaper when there's already one for the same screen
- [x]  test when the image doesn't exist
- [x] test garbage collection
- [x] update 05-default-rc-lua.lua
- [x] update the 03-declarative-widget "types of wiboxes" doc
- [ ] ~~add an example with @actionless hotkey widget (non-popup)~~
- [x] fix/port the xresources theme dynamic wallpaper
- [x] deprecate gears.wallpaper / fix tests/test-wallpaper
- [x] add a tests/test-awful-wallpaper.lua module

### CODE REVIEW

**Comment by actionless** [comment]
could we set new default svg wallpaper as one of the goals?

**Comment by Elv13** [comment]
image

Some progress on the doc examples. It doesn't pass the test and this is still pretty far from complete. The multi screen panning and more complex SVG corner cases (dpi, panning, stretching) are not handled yet (and the svg stretching wont be in the PR, this is already large enough).

@actionless Do you think you could refactor your hotkey popup to expose a real widget more cleanly? I would like a add some examples where the keybindings are part of the wallpaper itself. However, the current API doesn't make it easy to get the widget without creating the wibox.

**Comment by Aire-One** [comment]
At first, I was skeptical about the "use the widget API to draw wallpaper", now I see how good this is 😍

**Comment by actionless** [comment]
> Do you think you could refactor your hotkey popup to expose a real widget more cleanly?

not really, wibox-es themselves are already exposed in `_cached_wiboxes` property

and when you call `:show_help(c, s, show_args)`, based on function args it generates wibox with hotkeys and caching it inside aforementioned table

**Comment by actionless** [comment]
mb i could expose one of internal functions into api like `generate_hotkeys_layout(s, available_groups, show_awesome_keys)`?

**Comment by actionless** [comment]
ah, and few more things to take into consideration:

1) there could be so many hotkeys what they wouldnt fit into one page

2) pagination logic is relying on keygrabber

so without very clear goals is hard to say what and how should be changed

**Comment by Elv13** [comment]
I think it needs to be a widget that shows as many things as it can fit. Unless to want to use multi-screen to display multiple pages, which would be nice, there isn't much we can do when there isn't enough space.

So, at first, something that attempts to draw everything, then fail, is good enough. If something doesn't fit, the `wibox.layout.fixed.vertical/horizontal` wont draw them anyway. Maybe eventually we can add some magic to daisy chain a few with a `previous_page` property, but that's unnecessary for the "simple" use case?

**Comment by actionless** [comment]
i didnt got you clear - do you want a function for drawing a single page into a container widget?

**Comment by Elv13** [comment]
> i didnt got you clear - do you want a function for drawing a single page into a container widget?

I was thinking about exposing the hotkey popup as a widget (and by widget, I mean a wibox.widget instance, not a wibox instance with a widget in it. So, in practice, the `wibox.layout.fixed.something` instance as if there was infinite space available)

**Comment by actionless** [comment]
that would be too much to redo due to pagination,i wouldn't volunteer into that

File: lib/awful/wallpaper.lua [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+function module:get_dpi()
+    return self._private.context.dpi
+end
+
+function module:set_screen(s)
+    if not s then return end
+
+    self:_clear()
+    self:add_screen(s)
+    self:_update()
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:339]: `add_screen` already calls `_update`. I get it we have a mutex on the update method, but is it necessary to double call?

File: tests/examples/awful/template.lua [hunk 2]
```diff
@@ -58,11 +58,24 @@ if not rawget(screen, "no_outline") then
     end
 end
 
+-- Draw the wallpaper (if any).
+if root._wallpaper_pattern then
+    cr:set_source_rgb(1,0,0)
+    cr:set_line_width(2)
+    cr:rectangle(0, 0, root.size())
+    cr:stroke()
+    print("DRAW!!!!", root._wallpaper_pattern, root.size())
```
Comment (Aire-One) [inline tests/examples/awful/template.lua:67]: (note to not forget to remove this print)

**Review by Aire-One — COMMENTED** [review]
I'll try to play with this latter.

Do you have another available branch with more work on this field?

**Comment by Elv13** [comment]
> Do you have another available branch with more work on this field?

I don't. What I had/have is/was a textfile with the TODO. It now lives at the top of the PR with all the things to fix. I usually have branches with more advanced versions (usually in my `experimental` branch). But not for this. I made the awful_wallpaper branch in 2018!?! and just recently got back to it.

As it is pretty obvious by now, I am making little progress toward 4.4 with the current feature roadmap... Still, I am giving it a few hours a week. There's a lot of bugs to fix before I am ready to tag 4.4. All the refactoring work that has been merged since 4.3 has left quite a few scars in the code (like the notifications and keybinding rewrites). I wanted those projects to happen and I am happy they are merged, but they are causing some pain. In the grand scheme of things they were blocking my plans for 5.0. Obviously as it stands now, there is very little progress made on the larger 5.0 features ( `ruled.screen/tag/clipboard` rules API, built-in xrandr support, touchpad gestures, new titlebar API, new client layout API, etc).

**Comment by Elv13** [comment]
I finished the other half of this PR. I think technically, it is good enough for a final review and to merge. 

@actionless As mentioned in the first post, I added *some* DPI code, but it will still require some very serious changes to support multiple screens with multiple DPI and a single wallpaper. This will be for another PR, but at least the ground work is here

image

**Comment by actionless** [comment]
unfortunately i dont have now multiscreen configuration to test if same tiled pattern would look consistent on different-dpi displays, but overall docs looking very nice, so if there are any problems - they could be fixed in follow-up PRs

**Review by actionless — DISMISSED** [review]

**Comment by Elv13** [comment]
> if same tiled pattern would look consistent on different-dpi displays

For bitmap, it doesn't yet. There is still 2 properties to add. One to render larger section for each DPI and assemble them. This wont always work because a lot of widget don't scale with a "correct" pixel ratio. The other property is to either downscale or upscale the content after rendering it with the "highest" or "lowest" DPI. Those will look blurry, but at least align properly.

**Comment by Elv13** [comment]
Rebase to have the new Github Action fix.

Also add a few more HiDPI/SVG_wallpaper related commits.

image
image

@actionless With these commits, your `xresources` theme now use a `.svg` wallpaper with a CSS stylesheet to set the `xrdb` colors instead of using Lua code. There is some little regressions:

1) If librsvg isn't installed, it looks blurry
2) If the aspect ratio isn't 16:9, it gets centered. But since there is a fallback background color, it still look mostly the same.

**Comment by actionless** [comment]
cool! have you also tried that rich-gradient wallpaper from https://github.com/awesomeWM/awesome/pull/2867#issuecomment-538101962?

**Review by actionless — APPROVED** [review]

File: themes/xresources/theme.lua [hunk 3]
```diff
@@ -126,10 +126,38 @@ local wallpaper_alt_fg = xrdb.color12
 if not is_dark_bg then
     wallpaper_bg, wallpaper_fg = wallpaper_fg, wallpaper_bg
 end
-theme.wallpaper = function(s)
-    return theme_assets.wallpaper(wallpaper_bg, wallpaper_fg, wallpaper_alt_fg, s)
+
+local rsvg = pcall(function() return require("lgi").Rsvg end)
+
+if rsvg then
```
Comment (actionless) [inline themes/xresources/theme.lua:132]: > btw it should be a warning message instead of just sillently changing the behavior depending on presence of rsvg

**Review by actionless — CHANGES_REQUESTED** [review]
just noticed a minor problem while discussing another more abstract question

**Comment by Elv13** [comment]
> cool! have you also tried that rich-gradient wallpaper from #2867 (comment)?

Yes:

image

However I am not sure about the aspect ratio. I guess we should use 16:9 by default? Is there a way to make SVG smarter about aspect ratio? You know more about CSS and HTML than I. Could you research it?

> btw it should be a warning message instead of just sillently changing the behavior depending on presence of rsvg

I will publish an update shortly

File: themes/xresources/theme.lua (hunk 3 above)
Comment (Elv13) [reply themes/xresources/theme.lua:133]: done

**Comment by Elv13** [comment]
@actionless I addressed your comment for the last revision.

@Aire-One @sclu1034 Can you review this?

More generally, after 3 years of procrastination I finally completed the checklist from the first post. Over the last few revision I added more and more unit/doc/integration tests. This PR is now fully completed.

Future PRs could/will add more APIs to `gears.timer` to support "time of day" style signals to allow daytime and nighttime wallpapers. This requires a rather significant chunk of code and is too big for this PR. Another feature I found in other wallpaper tools is alternate patterns for the image tilling. Some allow diagonal lines or sine curves as an alternative to the grid pattern. I am still not sure what the API would look line in `wibox.widget.tile`.

As for HiDPI and SVG. There is still problems with `multi-screen` + `multi-dpi` wallpapers. More work will be needed as mentioned above fix the rendering and scaling. This PR is already big enough as it is. This is just too much.

Finally, one obvious and large "non-feature" is the lack of PyWall style ways to generate a theme from a wallpaper color palette. This is now "a thing" outside of `r/unixporn` too. The new version of Android has it built-in. This cannot be implemented without first adding proper support for `gears.thread`. The pixel level math used to generate those palette is too slow for the main thread. I don't plan to support it in the short/mid term. This is a v4.5 at the minimum, but probably even something like a v5.1 feature.

The test are flacky and fail in 

```lua
 - /home/runner/work/awesome/awesome/tests/test-awful-client.lua: Error: timeout waiting for signal in step 11/138 (@20) (...ner/work/awesome/awesome/tests/test-awful-client.lua:221).
```

This doesn't seem related to this PR. It looks like the Azure host is too slow or something. This will need debugging eventually.
```

**Comment by actionless** [comment]
> Finally, one obvious and large "non-feature" is the lack of PyWall style ways to generate a theme from a wallpaper color palette
> This cannot be implemented without first adding proper support for gears.thread

i wouldn't say that much of uninque pywal feature - i think first well-known implementation was done guy coleifer from reddit, slow and buggy alpha plugin for oomox (which was taking mb 2 minutes to generate the palette) and only few years after pywal

as to slow and buggy version from oomox - i got refactored it couple of years ago - and now it's not only in one file but also takes much faster (depends on resolution, if pic would be downscaled enough that would be really miliseconds) - so i can port that code from PIL + python to Cairo + lua (you can test the color-matching result in the oomox app itself, and also could compare it with few other libraries which are supported there, but porting them to lua would be harder as i'm not familiar with their codebase)

also one big difference between oomox algorhythm for color-extracting in compare to most of others - what it allows to specify reference color-palette: for example you specify what you want black, red light green and dark green - function return the most dominant black, red, light green and dark green colors

as regarding to thread thing - couldn't we do just hack with async_spawn-ing one more lua interpreter and writing result to temporary file? at least til proper threading stuff is there

**Review by actionless — APPROVED** [review]
and yup, i see the rsvg think is now handled

**Comment by Aire-One** [comment]
Hello @Elv13, I was busy IRL for the last week or so.

I'm glad to see there have been a lot of work done while I was away. Expect my reviews/comments/updates on currents PRs/issues as soon as I can catch up with them. :smiley:

File: tests/examples/awful/wallpaper/add_screen3.lua [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                    align  = "center",
+                    widget = wibox.widget.textbox,
+                },
+                widget = wibox.layout.stack
+            }
+        }
+
+--DOC_HIDE_END
+
+   wall:add_screen(screen[3])
```
Comment (Aire-One) [inline tests/examples/awful/wallpaper/add_screen3.lua:51]: Having only this line in the usage section sounds a bit confusing. I have no idea how to make it more clear, tho... Maybe this should be entirely hidden behind a --DOC_HIDE_ALL?

File: tests/examples/awful/wallpaper/corner1.lua [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    placement = require("awful.placement"), --DOC_HIDE
+} --DOC_HIDE
+local wibox = require("wibox") --DOC_HIDE
+local beautiful = require("beautiful") --DOC_HIDE
+
+screen[1]._resize {x = 0, y = 0, width = 320, height = 196} --DOC_HIDE
+
+
+   awful.wallpaper {
+       screen = screen[1],
```
Comment (Aire-One) [inline tests/examples/awful/wallpaper/corner1.lua:13]: Maybe it would be a good practice to first define the screen with something like `local s = screen[1] --DOC_HIDE` for example with only 1 screen. It would make the code more relatable to the user rc.lua and allow the code to work out of the box on multi monitor setups.

File: tests/examples/awful/wallpaper/tiled2.lua [hunk 6]
```diff
@@ -0,0 +1,30 @@
+--DOC_NO_USAGE --DOC_GEN_IMAGE
```
Comment (Aire-One) [inline tests/examples/awful/wallpaper/tiled2.lua:1]: Maybe it's too late right now for my brain to understand this example... What did you do to make the centered star be white while others are black?

File: lib/awful/wallpaper.lua [hunk 7]
```diff
@@ -0,0 +1,841 @@
...
+
+    self:repaint()
+end
+
+--- Detach the wallpaper from all screens.
+--
+-- Adding a new wallpaper to a screen will automatically
+-- detach the older one. However there is some case when
+-- it is useful to call this manually. For example, when
+-- adding a ned panned wallpaper, it is possible that 2
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:738]:
```suggestion
-- adding a new panned wallpaper, it is possible that 2
```
?

**Review by Aire-One — COMMENTED** [review]
This is amazing work! I think I'll need a second review to look more in depth in the wallpaper module itself. These are a few comments I had while reading the other files. Feel free to dismiss any of them.

**Comment by vp1981** [comment]
Hello Elv13. Till this PR is not merged could you update the PR? Now `test/examples/shims/screen.lua` has changes proposed in the PR so it is not applied on top of HEAD.

File: tests/examples/awful/wallpaper/tiled2.lua (hunk 6 above)
Comment (Elv13) [reply tests/examples/awful/wallpaper/tiled2.lua:1]: whoops, that was a bug in the tile container I forgot to fix. I added a commit for it

File: lib/wibox/widget/imagebox.lua [hunk 8]
```diff
@@ -48,28 +48,127 @@ end
...
+local rsvg_handle_cache = setmetatable({}, { __mode = 'k' })
 
 ---Load rsvg handle form image file
 ---@tparam string file Path to svg file.
 ---@return Rsvg handle
 local function load_rsvg_handle(file)
     if not Rsvg then return end
 
-    local cache = rsvg_handle_cache[file]
+    local cache = (rsvg_handle_cache[file] or {})["handle"]
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/wibox/widget/imagebox.lua]: It fails when `[file]` returns `nil`

File: docs/05-awesomerc.md.lua [hunk 9]
```diff
@@ -72,6 +72,15 @@ variables such as `bg_normal`. To get a list of all official variables, see
...
 ]]
 
+sections.DOC_WALLPAPER = [[
+The AwesomeWM wallpaper module, `awful.wallpaper` support both per-screen wallpaper
+and wallpaper across multiple screens. In the default configuration, `request::wallpaper`
+is emitted everytime a screen is added, moved, resized or when the bars
+(`awful.wibar`) are moved.
+
+This is will suited for single-screen wallpaper. If you wish to use multi-screen wallpaper,
+it is better to create a global wallpaper object and edit it when the screen change.
```
[the remark this answers is not in the document]
Comment (Elv13) [reply docs/05-awesomerc.md.lua:82]: done

File: lib/wibox/widget/imagebox.lua [hunk 10]
```diff
@@ -48,28 +48,127 @@ end
...
+
     if cache then
-        return cache
+        return cache, rsvg_handle_cache[file]
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/wibox/widget/imagebox.lua:62]: done

File: lib/wibox/widget/imagebox.lua [hunk 11]
```diff
@@ -48,28 +48,127 @@ end
...
+        handle, err = Rsvg.Handle.new_from_data(file)
+    else
+        handle, err = Rsvg.Handle.new_from_file(file)
     end
 
-    local handle, err = Rsvg.Handle.new_from_file(file)
     if not err then
-        rsvg_handle_cache[file] = handle
-        return handle
+        rsvg_handle_cache[file] = rsvg_handle_cache[file] or {}
```
[the remark this answers is not in the document]
Comment (Elv13) [reply lib/wibox/widget/imagebox.lua:74]: I don't think this is a problem for now. i don't think the ability to use raw SVG code will be used a lot beyond the tests/doc_examples

File: tests/examples/awful/wallpaper/add_screen3.lua (hunk 4 above)
Comment (Elv13) [reply tests/examples/awful/wallpaper/add_screen3.lua:51]: done

File: tests/examples/awful/wallpaper/corner1.lua (hunk 5 above)
Comment (Elv13) [reply tests/examples/awful/wallpaper/corner1.lua:13]: done

**Comment by Elv13** [comment]
> Hello Elv13. Till this PR is not merged could you update the PR? Now test/examples/shims/screen.lua has changes proposed in the PR so it is not applied on top of HEAD.

Can you clarify? I just rebased it again on top of HEAD and it just worked. Do you have other unmerged PRs or patches applied?

**Comment by vp1981** [comment]
Hello Emmanuel (Elv13).

Now everything works fine, thank you!

File: lib/wibox/widget/imagebox.lua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Aire-One) [reply lib/wibox/widget/imagebox.lua:60]: Oh, yes! Good catch

File: lib/wibox/widget/imagebox.lua (hunk 11 above)
[the remark this answers is not in the document]
Comment (Aire-One) [reply lib/wibox/widget/imagebox.lua:75]: Ok, let's keep in mind this will need improvement "later".

File: tests/examples/awful/wallpaper/tiled2.lua (hunk 6 above)
Comment (Aire-One) [reply tests/examples/awful/wallpaper/tiled2.lua:1]: I feel a little sad it's gone... It was so cool!

File: tests/examples/awful/wallpaper/add_screen3.lua (hunk 4 above)
Comment (Aire-One) [reply tests/examples/awful/wallpaper/add_screen3.lua:49]: I was more thinking of :

```lua
local s = screen[1] --DOC_HIDE

-- Then in the wallpaper constructor:
awful.wallpaper {
    screen = s,
    -- ...
}
```

This way the code in the example can be copied/pasted directly, as we consider (from the default awesomerc.lua) that the "request::wallpaper" callback have the screen as first parameter named s.

File: lib/awful/wallpaper.lua [hunk 12]
```diff
@@ -0,0 +1,841 @@
...
+    mutex = true
+
+    gtimer.delayed_call(function()
+        -- Remove the mutex first in case `paint()` raises an exception.
+        mutex = false
+        paint()
+    end)
+end
+
+capi.screen.connect_signal("removed", function(s)
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:397]: ~~Shouldn't we also do `background[s] = nil` in the scope of this function to garbage collect it? (and probably also do it for the pending_update table)~~ Actually, it's part of the remove_screen method...

File: lib/awful/wallpaper.lua [hunk 13]
```diff
@@ -0,0 +1,841 @@
...
+            table.remove(self._private.screens, k)
+        end
+    end
+
+    backgrounds[s] = nil
+
+    self:repaint()
+end
+
+local function new(_, args)
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:809]: Did you voluntary not document the module constructor, or did you just forget?

I think it's important to have this function documented, even if the module documentation is very descriptive about all usages. At least to have a full list of allowed arguments and types.

Comment (Elv13) [reply lib/awful/wallpaper.lua:809]: I forgot. I added it.

File: tests/examples/awful/wallpaper/add_screen3.lua (hunk 4 above)
Comment (Elv13) [reply tests/examples/awful/wallpaper/add_screen3.lua:49]: Well, this is the `add_screen` example. So I think it makes sense to use `add_screen`?

Comment (Aire-One) [reply tests/examples/awful/wallpaper/add_screen3.lua:49]: Oh wait, I replied on the wrong thread... Sorry

File: lib/awful/wallpaper.lua [hunk 14]
```diff
@@ -0,0 +1,857 @@
...
+            table.remove(self._private.screens, k)
+        end
+    end
+
+    backgrounds[s] = nil
+
+    self:repaint()
+end
+
+--- Create a wallpaper.
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:809]:
```suggestion
--- Create a wallpaper.
--
-- Note that all parameters are not required. Please refer to the module description and examples to understand parameters usages.
```

File: lib/awful/wallpaper.lua [hunk 15]
```diff
@@ -0,0 +1,857 @@
...
+-- @tparam number args.dpi The wallpaper DPI (dots per inch).
+-- @tparam screen args.screen The wallpaper screen.
+-- @tparam table args.screens A list of screen for this wallpaper.
+-- @tparam gears.color args.bg The background color.
+-- @tparam gears.color args.fg The foreground color.
+-- @tparam gears.color args.uncovered_areas_color The color for the uncovered areas.
+-- @tparam boolean args.honor_workarea Honor the workarea.
+-- @tparam boolean args.honor_padding Honor the screen padding.
+-- @tparam table args.uncovered_areas Returns the list of screen(s) area which won't be covered by the wallpaper.
+-- @tparam function|string args.panning_area Defines where the wallpaper is placed when there is multiple screens.
```
Comment (Aire-One) [inline lib/awful/wallpaper.lua:823]:
```suggestion
-- @constructorfct awful.wallpaper
-- @tparam table args
-- @tparam[opt] wibox.widget args.widget The wallpaper widget.
-- @tparam[opt] number args.dpi The wallpaper DPI (dots per inch).
-- @tparam[opt] screen args.screen The wallpaper screen.
-- @tparam[opt] table args.screens A list of screen for this wallpaper. Use this parameter as a remplacement for `args.screen` to manage multiscreen wallpaper. (Note: the expected table should be an array-like table `{screen1, screen2, ...}`)
-- @tparam[opt] gears.color args.bg The background color.
-- @tparam[opt] gears.color args.fg The foreground color.
-- @tparam[opt] gears.color args.uncovered_areas_color The color for the uncovered areas.
-- @tparam[opt] boolean args.honor_workarea Honor the workarea.
-- @tparam[opt] boolean args.honor_padding Honor the screen padding.
-- @tparam[opt] table args.uncovered_areas Returns the list of screen(s) area which won't be covered by the wallpaper.
-- @tparam[opt] function|string args.panning_area Defines where the wallpaper is placed when there is multiple screens.
```

**Review by Aire-One — COMMENTED** [review]
This is what I think to be better practices for constructor references doc. As always, feel free to dismiss suggestions.

**Comment by Aire-One** [comment]
To clarify my idea with the screen parameter in example :

It was a "general" remark, and wasn't meat to be specifically in the add_scree example (especially because it's the place where this suggestion would make nonsense).

Following my suggestion to not use `screen = screen[1]` in all examples, you have "removed" all of them with `--DOC_HIDE`. 

Let's take the very first example as it is rendered in the doc :

```lua
awful.wallpaper {
    widget = {
        {
            image  = beautiful.wallpaper,
            resize = true,
            widget = wibox.widget.imagebox,
        },
        valign = "center",
        halign = "center",
        tiled  = false,
        widget = wibox.container.tile,
    }
}
```

My first idea was to keep the screen parameter displayed, and show it as `screen = s` to mimic what the default awesomerc.lua would have.
We could have done that with something like this in the example's files :

```lua
local s = screen[1] --DOC_HIDE

-- Then in the wallpaper constructor:
awful.wallpaper {
    screen = s,
    -- ...
}
```

However, I think I have mixed opinion about it now... Maybe it overcrowds the usage example, and it's better to keep it hidden anyway? :shrug:

**Review by Aire-One — DISMISSED** [review]
I have this PR merged and used in my local awesome config for about 1 week now.

This is amazing work @Elv13! :rocket:

**Comment by actionless** [comment]
however, i see more of github-actions setups are failing than usually (usually only the one with coveralls failing due to tokens' thing)

**Comment by Elv13** [comment]
> however, i see more of github-actions setups are failing than usually (usually only the one with coveralls failing due to tokens' thing)

Yeah, there was a `luacheck` issue with a line being too long in the last revision. This is fixed now.

Soooo, last call, I got 2 approvals of 99.99995% the same code as the current revision. I will merge this PR tomorrow if there is no new findings.


As for what's next. First of all, the point of doing this before v4.4 was/is for the SVG wallpaper support. So this will get unblocked. Also, the "spiritual successor" of this PR is a large rework of `gears.timer` to:

* allow better slideshows for the wallpaper module
* Add daytime/nighttime wallpaper example
* add a "do not disturb" style examples to `naughty.suspended`
* fix some `textclock` bugs marked for v4.4 (caused by `gears.timer`)
* fix laptop related bugs when suspending/resuming, also marked for v4.4 (mitigated using `gears.timer`)
* implement some old feature requests I got from IRC/Reddit over the years
* documentation and examples, as usual when I revisit old modules

This is actually a rather large PR, so I am not sure when I will finish it. It's progressing well. I also have a couple more v4.4 bug fixes PRs coming "soon". I would still appreciate of some people would step up and help fix the bugs marked for v4.4 milestone.

I will also bring the old `gears.watcher` commits into that PR. I have yet to finish the changes @actionless asked for `gears.reactive` and `gears.connection`. They obviously wont make it for 4.4 given how much work it would be to do the linting script. However, `gears.watcher` is "fine" and not affected by those concerns.

At this rate I/we could try to aims to make the deadline for Ubuntu 22.04 for the v4.4 release. Which means it has to be out by Xmas. Still, there's still a hole lo of issues marked for v4.4
